### ISSUE TITLE
EmptyDataProvider

### ISSUE THREAD

**Issue by leandrogehlen** [yiisoft/yii2#6373]
there are situations, where it is necessary, to show, initially, empty datagrid . 
After the user to select required filters, then it will execute query.

Currently, it is necessary to execute a query that not return data.

I don't think it is necessary to execute this query,  knowing that any data it will returned?

How can we solve this problem?

**Comment by SilverFire** [comment]
Resolved by https://github.com/yiisoft/yii2/commit/53e1018648af6b79d429687ce685c5d7eb6fd054

**Comment by leandrogehlen** [comment]
The label problems was solved, but the how to show,  initially, empty datagrid whit any dataprovder?

**Comment by SilverFire** [comment]
Use the approach suggested by @samdark

**Comment by leandrogehlen** [comment]
It's working, but the issue is trigger query to show empty grid.
Currently, is necessary to write something like: 
```php
$query = Movfinan::find()->where(['id' => null]); // this is not necessary, but, currently, it's the unique way

$dataProvider = new ActiveDataProvider([
     'query' => $query     
]);

```

**Comment by SilverFire** [comment]
Well yeah...

**Comment by Faryshta** [comment]
what do you suggest @leandrogehlen ?

**Comment by leandrogehlen** [comment]
@cebe  suggested create a method: 

```php
public function forceEmpty()
{
    $this->setModels([]); 
    $this->setKeys([]); 
    $this->setTotalcount(0);
}
``` 
But this is the solution to ArrayDataProvider.

i think better to add `autoload` property into ```BaseDataProvider```, then all provider can have this feature

**Comment by Faryshta** [comment]
can you make a pr?

**Comment by leandrogehlen** [comment]
Yes, but today, i can't.
Could be tomorow?

**Comment by Faryshta** [comment]
@leandrogehlen yup, its not obligatory :+1:

**Comment by leandrogehlen** [comment]
I was trying to implement this feature, but i think that the implementation made by @SilverFire  resolved the issue.

```php
public function actionIndex()
{
    if (Yii::$app->request->get('empty', false)) {
        $dataProvider = new ArrayDataProvider([
            'modelClass' => Person::className()
        ]); 
    } else {
        $query = Person::find();
        $dataProvider = new ActiveDataProvider([
            'query' => $query
        ])
    }      

    return $this->render('index', [
        'dataProvider' => $dataProvider
    ]);
}

```

**Comment by Faryshta** [comment]
@leandrogehlen how about doing that on the `ModelSearch::search()` method?

you can count the number of results with `$query->count()` if its zero, then return the `ArrayDataProvider` else the `ActiveDataProvider`

**Comment by leandrogehlen** [comment]
But the intention is to avoid query trigger.
Currently, when result of query is empty, this featrue is working

**Comment by klimov-paul** [comment]
Relates to #12390

Perhaps there should be an option inside `yii\db\Query` which can turn off the actual query running as it will not return any result anyway instead of using `1=0` condition.

**Comment by klimov-paul** [comment]
Possible fix: #12708

**Comment by leandrogehlen** [comment]
I think is an excelent solution
